### ISSUE TITLE
Exception when trying to update cart with an item which is OOS

### ISSUE THREAD

**Issue by nikunjkotecha** [acquia/commerce-manager-magento#69]
# Scenario
I visited site on Sunday (today is Monday) when there was a sale going on, I added 3 items to Cart and for some reason didn't finish checkout. I come back on site today and when I visit Cart page I see 2 out of the 3 items are OOS now. I try to remove OOS ones from Cart but I'm not able to do.

# Background
We have single API for Update Cart, when user tries to add/update/remove an item in Cart we update all the items and if any one item is OOS or invalid we throw an exception. This doesn't allow
- adding an item when another item in cart is OOS
- removing an item if another item in cart is OOS (even if we are trying to remove an OOS item itself or another item which is in stock)
- basically for the end user cart is locked and he/she can't do anything

# Suggested solution
We split the update cart processing into multiple actions
- add item
- update item
- remove item
- apply coupon
- remove coupon
- add shipping info
- remove shipping info
- add/update billing info
- add/update payment info

On Drupal side, split the functions and send only required info with the additional info about the requested action

This in theory will allow us to mimic what exactly happens in MDC FE and will allow us to manage add/update/remove in better way. We can still validate cart's items when actions related to shipping/billing/payment are performed.

### Additional update:
We handle all exceptions in better way, respond back with exception in a response itself with proper message instead of throwing 400 or 500 error. This is already done today for apply coupon.

**Comment by nikunjkotecha** [comment]
cc @sdelbosc @vbouchet31 @malachy-mcconnell

**Comment by sdelbosc** [comment]
@malachy-mcconnell if you have few mins could you just read this one and share thoughts?

**Comment by malachy-mcconnell** [comment]
I agree with Nikunj. I have long had concerns over the cart handling in the Magneto module. It is always best to use the Magento processes. The idea of just assigning all the cart products in one go always made me uneasy but it seemed to work. Here is a sort-of edge case but actually it exposes that we are not following correct Magento way and I would welcome a ticket to review and overhaul the ACM Magento module's cart process.

**Comment by sdelbosc** [comment]
Thanks a lot @malachy-mcconnell.
